### PR TITLE
Add documentation for `CodeGenerator`s.

### PR DESCRIPTION
**Pull request by vRallev**

### CODE REVIEW

File: compiler-api/README.md [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+}
+```
+
+Note that the sample code generator is annotated with `@AutoService`. It's recommended to use the
+[AutoService](https://github.com/google/auto/tree/master/service) library to generate necessary 
+code for `ServiceLoader` API. Anvil uses this mechanism to load your custom code generator.
+
+You can generate as many classes and files as you wish. You can even generate code that uses Anvil
+or Dagger annotations and Anvil will process these files. The `generateCode()` function is called 
+multiple times until no code generators generated any code anymore.
```
Comment (lmchew) [inline compiler-api/README.md:66]: Minor correction: "... until no code generators generate code anymore."

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,5 +1,11 @@
 # Changelog
 
+## 2.3.0
+
+* Add option to extend Anvil with your own `CodeGenerator`, see [here](README.md#extending-anvil) and #265.
+* Use Gradle Property APIs in the Anvil extension. This is a breaking change for Kotlin and .kts consumers of the Anvil Gradle plugin, see #284.
```
Comment (ZacSweers) [inline CHANGELOG.md:6]:
```suggestion
* Use Gradle Property APIs in the Anvil extension. This is a source-breaking change (but binary-compatible) for Kotlin and .kts consumers of the Anvil Gradle plugin, see #284.
```

File: README.md [hunk 3]
```diff
@@ -201,7 +201,48 @@ wrong imports and deeply nested dependency chains applications might need to mak
 exclusion rule does what it implies. In this specific example `DaggerModule` wishes to be
 contributed to this scope, but it has been excluded for this component and thus is not added.
 
-## Advantages
+## Dagger Factory generation
```
Comment (ZacSweers) [inline README.md:204]:
```suggestion
## Dagger Factory Generation
```

Either cap them all or only capitalize Dagger

File: compiler-api/README.md [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+
+To use your new code generator in any module you need to add the module to Anvil classpath:
```
Comment (lmchew) [inline compiler-api/README.md:68]: I think this should be "... to Anvil's classpath:" or maybe "... to the Anvil classpath:".

File: README.md [hunk 5]
```diff
@@ -201,7 +201,48 @@ wrong imports and deeply nested dependency chains applications might need to mak
...
+Anvil only processes Kotlin code, you shouldn't enable it in modules with mixed Kotlin / Java
+sources either.
+
+When you enable this feature, don't forget to remove the Dagger annotation processor. You should
+keep all other dependencies.
+
+## Extending Anvil
+
+Every codebase has its own dependency injection patterns where certain code structures need to be
+repeated over and over again. Here Anvil comes to the rescue, and you can extend the compiler
```
Comment (ZacSweers) [inline README.md:241]:
```suggestion
repeated over and over again. Here Anvil comes to the rescue and you can extend the compiler 
```

File: README.md [hunk 6]
```diff
@@ -201,7 +201,48 @@ wrong imports and deeply nested dependency chains applications might need to mak
...
+sources either.
+
+When you enable this feature, don't forget to remove the Dagger annotation processor. You should
+keep all other dependencies.
+
+## Extending Anvil
+
+Every codebase has its own dependency injection patterns where certain code structures need to be
+repeated over and over again. Here Anvil comes to the rescue, and you can extend the compiler 
+plugin with your own code generator. For steps please take a look at the
```
Comment (ZacSweers) [inline README.md:242]:
```suggestion
plugin with your own `CodeGenerator`. For usage please take a look at the 
```

File: README.md [hunk 7]
```diff
@@ -236,6 +277,9 @@ compile tasks are skipped entirely, if nothing has changed. This doesn't change
 
 ![Benchmark](images/benchmark.png?raw=true "Benchmark")
 
+On top of that Anvil provides actual build time improvements by replacing the Dagger annotation
```
Comment (ZacSweers) [inline README.md:280]:
```suggestion
On top of that, Anvil provides actual build time improvements by replacing the Dagger annotation
```

File: compiler-api/README.md [hunk 8]
```diff
@@ -0,0 +1,86 @@
+# Compiler-API
+
+Anvil is composed of several independent code generators that provide hints for the later stages
+in the build graph when all modules and components are merged together. These code generators 
+are similar to annotation processors in the way that they look for a specific annotation like 
+`@ContributesTo` and then generate necessary code. 
+
+If you see yourself repeating the same code structures and patterns for your dependency injection
+setup over and over again, then you can extend Anvil and implement your own code generator.
```
Comment (ZacSweers) [inline compiler-api/README.md:9]:
```suggestion
setup over and over again, then you can extend Anvil and implement your own code generator via the `CodeGenerator` interface.
```

**Review by ZacSweers — APPROVED** [review]

**Review by lmchew — APPROVED** [review]

**Pull request merged**